File: Shared/Networking/SofseDTOs.swift
import Foundation

// DTOs internos de decodificación. Cubren los dos esquemas de /arribos (vivo y horario)
// y las alertas embebidas en /infraestructura.

// Valor JSON de tipo desconocido. Solo sirve para detectar presencia (ej: cancelacion).
struct AnyJSON: Decodable {
    init(from decoder: Decoder) throws {
        if let c = try? decoder.singleValueContainer() { _ = c.decodeNil() }
    }
}

struct ArribosResponse: Decodable {
    let timestamp: Int?
    let results: [ResultDTO]?
    let total: Int?
}

struct ResultDTO: Decodable {
    let arribo: ArriboDTO?
    let servicio: ServicioDTO?
}

struct ArriboDTO: Decodable {
    let orden: Int?
    let nombre: String?
    let idElemento: Int?
    let segundos: Int?
    let anden: AndenDTO?
    let equipo: EquipoDTO?
    let llegada: TimingDTO?
    let salida: TimingDTO?
}

struct AndenDTO: Decodable {
    let id: Int?
    let nombre: String?
}

struct EquipoDTO: Decodable {
    let id: Int?
    let nombre: String?
    let esElectrico: Int?
    let gpss: [String]?
}

struct TimingDTO: Decodable {
    let programada: String?
    let estimada: String?
    let real: String?
    let latitud: Double?
    let longitud: Double?
}

struct LocationDTO: Decodable {
    let lat: Double?
    let long: Double?
}

struct GerenciaRefDTO: Decodable {
    let id: Int?
    let nombre: String?
}

struct EstadoDTO: Decodable {
    let id: Int?
    let nombre: String?
}

struct DesdeDTO: Decodable {
    let estado: EstadoDTO?
}

// Cabecera de ramal. Maneja nombreCorto (vivo) y nombre_corto (horario).
struct CabeceraDTO: Decodable {
    let id: Int?
    let nombre: String?
    let nombreCorto: String?

    enum K: String, CodingKey { case id, nombre, nombreCorto, nombre_corto }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: K.self)
        id = try? c.decodeIfPresent(Int.self, forKey: .id) ?? nil
        nombre = try? c.decodeIfPresent(String.self, forKey: .nombre) ?? nil
        let camel = (try? c.decodeIfPresent(String.self, forKey: .nombreCorto) ?? nil) ?? nil
        let snake = (try? c.decodeIfPresent(String.self, forKey: .nombre_corto) ?? nil) ?? nil
        nombreCorto = camel ?? snake
    }

    var display: String { nombreCorto ?? nombre ?? "" }
}

struct RamalRefDTO: Decodable {
    let id: Int?
    let nombre: String?
    let siglas: String?
    let tolerancia: Int?
    let cabeceraInicial: CabeceraDTO?
    let cabeceraFinal: CabeceraDTO?
}

struct EstacionRecorridoDTO: Decodable {
    let idElemento: Int?
    let nombre: String?
    let orden: Int?
    let anden: AndenDTO?
    let llegada: TimingDTO?
    let salida: TimingDTO?
}

struct ServicioDTO: Decodable {
    let id: String?
    let numero: Int?
    let sentido: Int?
    let cancelacion: AnyJSON?
    let location: LocationDTO?
    let gerencia: GerenciaRefDTO?
    let ramal: RamalRefDTO?
    let desde: DesdeDTO?
    let estaciones: [EstacionRecorridoDTO]?
}

// --- Alertas ---

struct GerenciaAlertDTO: Decodable {
    let id: Int?
    let alerta: [AlertaDTO]?
}

struct RamalAlertDTO: Decodable {
    let id: Int?
    let alerta: AlertaDTO?
}

struct AlertaDTO: Decodable {
    let id: Int?
    let linea_id: Int?
    let ramal_id: Int?
    let causa_gtfs: String?
    let efecto_gtfs: String?
    let contenido: String?
    let criticidad_orden: Int?
    let vigencia_desde: String?
    let vigencia_hasta: String?
}

extension ServiceAlert {
    init(dto: AlertaDTO) {
        self.init(
            id: dto.id ?? abs((dto.contenido ?? UUID().uuidString).hashValue),
            lineId: dto.linea_id,
            ramalId: dto.ramal_id,
            causeGTFS: dto.causa_gtfs ?? "OTHER_CAUSE",
            effectGTFS: dto.efecto_gtfs ?? "OTHER_EFFECT",
            content: dto.contenido ?? "",
            criticality: dto.criticidad_orden ?? 99,
            validFrom: SofseDate.alert(dto.vigencia_desde),
            validUntil: SofseDate.alert(dto.vigencia_hasta)
        )
    }
}

File: android/app/src/main/java/com/alandaitch/anden/data/net/SofseDtos.kt
package com.alandaitch.anden.data.net

import com.alandaitch.anden.data.model.ServiceAlert
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNames
import kotlinx.serialization.json.JsonNull

// DTOs internos de decodificación. Cubren los dos esquemas de /arribos (vivo y horario)
// y las alertas embebidas en /infraestructura.

@Serializable
data class ArribosResponse(
    val timestamp: Long? = null,
    val results: List<ResultDTO>? = null,
    val total: Int? = null
)

@Serializable
data class ResultDTO(
    val arribo: ArriboDTO? = null,
    val servicio: ServicioDTO? = null
)

@Serializable
data class ArriboDTO(
    val orden: Int? = null,
    val nombre: String? = null,
    val idElemento: Int? = null,
    val segundos: Int? = null,
    val anden: AndenDTO? = null,
    val equipo: EquipoDTO? = null,
    val llegada: TimingDTO? = null,
    val salida: TimingDTO? = null
)

@Serializable
data class AndenDTO(val id: Int? = null, val nombre: String? = null)

@Serializable
data class EquipoDTO(
    val id: Int? = null,
    val nombre: String? = null,
    val esElectrico: Int? = null,
    val gpss: List<String>? = null
)

@Serializable
data class TimingDTO(
    val programada: String? = null,
    val estimada: String? = null,
    val real: String? = null,
    val latitud: Double? = null,
    val longitud: Double? = null
)

@Serializable
data class LocationDTO(val lat: Double? = null, val long: Double? = null)

@Serializable
data class GerenciaRefDTO(val id: Int? = null, val nombre: String? = null)

@Serializable
data class EstadoDTO(val id: Int? = null, val nombre: String? = null)

@Serializable
data class DesdeDTO(val estado: EstadoDTO? = null)

// Cabecera de ramal. Maneja nombreCorto (vivo) y nombre_corto (horario) vía JsonNames.
@OptIn(ExperimentalSerializationApi::class)
@Serializable
data class CabeceraDTO(
    val id: Int? = null,
    val nombre: String? = null,
    @JsonNames("nombreCorto", "nombre_corto") val nombreCorto: String? = null
) {
    val display: String get() = nombreCorto ?: nombre ?: ""
}

@Serializable
data class RamalRefDTO(
    val id: Int? = null,
    val nombre: String? = null,
    val siglas: String? = null,
    val tolerancia: Int? = null,
    val cabeceraInicial: CabeceraDTO? = null,
    val cabeceraFinal: CabeceraDTO? = null
)

@Serializable
data class EstacionRecorridoDTO(
    val idElemento: Int? = null,
    val nombre: String? = null,
    val orden: Int? = null,
    val anden: AndenDTO? = null,
    val llegada: TimingDTO? = null,
    val salida: TimingDTO? = null
)

@Serializable
data class ServicioDTO(
    val id: String? = null,
    val numero: Int? = null,
    val sentido: Int? = null,
    val cancelacion: JsonElement? = null,
    val location: LocationDTO? = null,
    val gerencia: GerenciaRefDTO? = null,
    val ramal: RamalRefDTO? = null,
    val desde: DesdeDTO? = null,
    val estaciones: List<EstacionRecorridoDTO>? = null
) {
    // Cancelación presente = cualquier valor no nulo (ni JSON null).
    val isCancelled: Boolean get() = cancelacion != null && cancelacion !is JsonNull
}

// --- Alertas ---

@Serializable
data class GerenciaAlertDTO(val id: Int? = null, val alerta: List<AlertaDTO>? = null)

@Serializable
data class RamalAlertDTO(val id: Int? = null, val alerta: AlertaDTO? = null)

@Serializable
data class AlertaDTO(
    val id: Int? = null,
    val linea_id: Int? = null,
    val ramal_id: Int? = null,
    val causa_gtfs: String? = null,
    val efecto_gtfs: String? = null,
    val contenido: String? = null,
    val criticidad_orden: Int? = null,
    val vigencia_desde: String? = null,
    val vigencia_hasta: String? = null
)

// Convierte el DTO de alerta al modelo de dominio.
fun AlertaDTO.toServiceAlert(): ServiceAlert = ServiceAlert(
    id = id ?: kotlin.math.abs((contenido ?: java.util.UUID.randomUUID().toString()).hashCode()),
    lineId = linea_id,
    ramalId = ramal_id,
    causeGTFS = causa_gtfs ?: "OTHER_CAUSE",
    effectGTFS = efecto_gtfs ?: "OTHER_EFFECT",
    content = contenido ?: "",
    criticality = criticidad_orden ?: 99,
    validFrom = SofseDate.alert(vigencia_desde),
    validUntil = SofseDate.alert(vigencia_hasta)
)
